[Exercise 5.15] Fix test case

xuebingli opened the pull request:
`tails` should return an empty list. Thus `Stream.of(1, 2, 3).tails()` should include `Stream.empty()`.

xuebingli commented:
cc @marc0der

marc0der commented:
How about adding an `append` function to `Stream` to make this look a bit more elegant?

```
    fun <A> Stream<A>.tails(): Stream<Stream<A>> =
        unfold(this) { s: Stream<A> ->
            when (s) {
                is Cons ->
                    Some(Pair(s, s.tail()))
                else -> None
            }
        }.append(Stream.empty<A>())
```

xuebingli commented:
I've refactored with `append` as suggested. A couple of things.
1. `append` accepts `() -> Stream<A>` instead of `Stream<A>` as argument, thus the use of `{` and `}`.
2. We need to pass `{ Stream.empty<Stream<A>>() }` instead of `{ Stream.empty<A>() }` to `append`, thus I've removed the type signature and used `{ Stream.empty() }` for succinctness.
3. `append` in Solution 5.7 could not be imported, thus I moved the function to the package level. I did not move all other functions due to a name conflict - `map` is already defined in Solution 5.13.

marc0der commented:
Thanks for this fix, looks great!

kalistace commented:
Odd, this change fails for me:

```
Expected :Cons(head=Cons(head=1, tail=Cons(head=2, tail=Cons(head=3, tail=chapter3.Nil@50b37555))), tail=Cons(head=Cons(head=2, tail=Cons(head=3, tail=chapter3.Nil@50b37555)), tail=Cons(head=Cons(head=3, tail=chapter3.Nil@50b37555), tail=Cons(head=chapter3.Nil@50b37555, tail=chapter3.Nil@50b37555))))
Actual   :Cons(head=Cons(head=1, tail=Cons(head=2, tail=Cons(head=3, tail=chapter3.Nil@50b37555))), tail=Cons(head=Cons(head=2, tail=Cons(head=3, tail=chapter3.Nil@50b37555)), tail=Cons(head=Cons(head=3, tail=chapter3.Nil@50b37555), tail=chapter3.Nil@50b37555)))
```

Previous solution seems to work however :

 ```
     fun <A> Stream<A>.tails(): Stream<Stream<A>> =
        Stream.cons(
            { this },
            {
                unfold(this) { s: Stream<A> ->
                    when (s) {
                        is Cons ->
                            Some(Pair(s.tail(), s.tail()))
                        else -> None
                    }
                }

            }
        )
        ```
